components/layout_2020/table/layout.rs

```diff
@@ -918,88 +1007,356 @@ impl<'a> TableLayout<'a> {
+            // last one.
+            if rowspan_distribution && only_have_empty_rows {
+                track_sizes[*empty_rows.last().unwrap()] += excess_size;
+                return;
+            }
+
+            let non_empty_rows_all_constrained = non_empty_rows
+                .iter()
+                .filter(|index| is_unconstrained(*index))
+                .next()
+                .is_none();
+
+            // Otherwise, if we only have empty rows or if all the non-empty rows are constrained,
+            // then grow the empty rows.
+            if only_have_empty_rows || non_empty_rows_all_constrained {
+                // If there are both unconstrained and constrained empty rows, only increase the
+                // size of the unconstrained ones, otherwise increase the size of all empty rows.
+                let mut rows_to_grow = &empty_rows;
+                let unconstrained_empty_rows: Vec<usize> = rows_to_grow
+                    .iter()
+                    .map(|track| *track)
+                    .filter(is_unconstrained)
+                    .collect();
+                if !unconstrained_empty_rows.is_empty() {
+                    rows_to_grow = &unconstrained_empty_rows;
+                }
+
+                // All empty rows that will grow equally.
+                let scale = 1.0 / rows_to_grow.len() as f32;
+                for track_index in rows_to_grow.iter() {
+                    track_sizes[*track_index] += excess_size.scale_by(scale);
+                }
+                return;
+            }
+        }
 
-            self.row_baselines.push(max_ascent);
-            self.row_sizes
-                .push(max_row_height.max(max_ascent + max_descent));
+        // If there are non-empty rows, they all grow in proportion to their current size,
+        // whether or not they are constrained.
+        if !non_empty_rows.is_empty() {
+            let total_size: Au = non_empty_rows.iter().map(|index| track_sizes[*index]).sum();
+            for track_index in non_empty_rows.iter() {
+                let scale = track_sizes[*track_index].to_f32_px() / total_size.to_f32_px();
+                track_sizes[*track_index] += excess_size.scale_by(scale);
+            }
         }
     }
 
+    /// Given computed row sizes, compute the final block size of the table and distribute extra
+    /// block size to table rows.
+    fn compute_table_height_and_final_row_heights(
+        &mut self,
+        mut row_sizes: Vec<Au>,
+        containing_block: &ContainingBlock,
+    ) {
+        // The table content height is the maximum of the computed table height from style and the
+        // sum of computed row heights from row layout plus size from borders and spacing.
+        let table_height_from_style = self
+            .table
+            .style
+            .content_box_size(containing_block, &self.pbm)
+            .auto_is(Length::zero)
+            .block
```

Comment:
No need to resolve both dimensions:

```suggestion
            .block
            .auto_is(Length::zero)
```

Also this is not taking min/max-height into account, but we can leave it for the follow-up.